shared_model/backend/protobuf/common_objects/proto_common_objects_factory.hpp

```diff
@@ -0,0 +1,259 @@
+                         const interface::types::AssetIdType &asset_id,
+                         const interface::Amount &balance) override {
+        iroha::protocol::AccountAsset asset;
+        asset.set_account_id(account_id);
+        asset.set_asset_id(asset_id);
+        auto proto_balance = asset.mutable_balance();
+        convertToProtoAmount(*proto_balance->mutable_value(),
+                             balance.intValue());
+        proto_balance->set_precision(balance.precision());
+
+        auto proto_asset = std::make_unique<AccountAsset>(std::move(asset));
+
+        auto answer =
+            validate(*proto_asset, [this](const auto &asset, auto &reasons) {
+              validator_.validateAccountId(reasons, asset.accountId());
+              validator_.validateAssetId(reasons, asset.assetId());
+            });
+
+        if (answer) {
+          return iroha::expected::makeError(answer.reason());
+        }
+
+        return iroha::expected::makeValue<
+            std::unique_ptr<interface::AccountAsset>>(std::move(proto_asset));
+      }
+
+      FactoryResult<std::unique_ptr<interface::Amount>> createAmount(
+          boost::multiprecision::uint256_t value,
+          interface::types::PrecisionType precision) override {
+        iroha::protocol::Amount amount;
+        amount.set_precision(precision);
+        convertToProtoAmount(*amount.mutable_value(), value);
+
+        auto proto_amount = std::make_unique<Amount>(std::move(amount));
+
+        auto answer = validate(*proto_amount, [](const auto &, auto &) {
+          // no validation needed,
+          // since any amount is valid in general context
+        });
+
+        if (answer) {
+          return iroha::expected::makeError(answer.reason());
+        }
+
+        return iroha::expected::makeValue<std::unique_ptr<interface::Amount>>(
+            std::move(proto_amount));
+      }
+
+      FactoryResult<std::unique_ptr<interface::Amount>> createAmount(
+          std::string amount) override {
+        // taken from iroha::model::Amount
+        // check if valid number
+        const static std::regex e("([0-9]*\\.[0-9]+|[0-9]+)");
+        if (!std::regex_match(amount, e)) {
+          return iroha::expected::makeError("number string is invalid");
+        }
+
+        // get precision
+        auto dot_place = amount.find('.');
+        interface::types::PrecisionType precision;
+        if (dot_place > amount.size()) {
+          precision = 0;
+        } else {
+          precision = amount.size() - dot_place - 1;
+          // erase dot from the string
+          amount.erase(std::remove(amount.begin(), amount.end(), '.'),
+                       amount.end());
+        }
+
+        auto begin = amount.find_first_not_of('0');
+
+        // create uint256 value from obtained string
+        boost::multiprecision::uint256_t value = 0;
+        if (begin <= amount.size()) {
+          value = boost::multiprecision::uint256_t(amount.substr(begin));
```

Comment:
Isn't overflow possible anywhere? Of course, I don't think someone's going to provide more than uint256, but still, for safety reasons, is it possible to handle such kind of error?

Reply:
This code is taken from lib/amount. It has worked so far, and I do not want to touch it :). It will be changed in the future, then we can replace it here